src/format/mod.rs

```diff
@@ -0,0 +1,176 @@
+        .into_iter()
+        .map(|mismatch| {
+            let lines = mismatch.lines.iter();
+            let num_removed = lines
+                .filter(|line| match line {
+                    DiffLine::Resulting(_) => true,
+                    _ => false,
+                })
+                .count();
+
+            let new_lines: Vec<String> = mismatch
+                .lines
+                .into_iter()
+                .filter_map(|line| match line {
+                    DiffLine::Context(_) | DiffLine::Resulting(_) => {
+                        None
+                    }
+                    DiffLine::Expected(str) => Some(str),
+                })
+                .collect();
+            let start_line = mismatch.line_number_orig - 1;
+            let end_line = start_line + (num_removed as u32);
+
+            let new_text = new_lines.join("\n");
+
+            TextEdit {
+                range: Range {
+                    start: Position::new(start_line, 0),
+                    end: Position::new(end_line, 0),
+                },
+                new_text,
+            }
+        })
+        .collect()
+}
+
+#[derive(Debug, PartialEq)]
+enum DiffLine {
+    Context(String),
+    Expected(String),
+    Resulting(String),
+}
+
+#[derive(Debug, PartialEq)]
+struct Mismatch {
+    /// The line number in the formatted version.
+    pub line_number: u32,
+    /// The line number in the original version.
+    pub line_number_orig: u32,
+    /// The set of lines (context and old/new) in the mismatch.
+    pub lines: Vec<DiffLine>,
+}
+
+impl Mismatch {
+    fn new(line_number: u32, line_number_orig: u32) -> Mismatch {
+        Mismatch {
+            line_number,
+            line_number_orig,
+            lines: Vec::new(),
+        }
+    }
+}
+
+// Produces a diff between the expected output and actual output.
+fn make_diff(
+    expected: &str,
+    actual: &str,
+    context_size: usize,
+) -> Vec<Mismatch> {
+    let mut line_number = 1;
+    let mut line_number_orig = 1;
+    let mut context_queue: VecDeque<&str> =
+        VecDeque::with_capacity(context_size);
+    let mut lines_since_mismatch = context_size + 1;
+    let mut results = Vec::new();
+    let mut mismatch = Mismatch::new(0, 0);
+
+    for result in diff::lines(expected, actual) {
+        match result {
+            diff::Result::Left(str) => {
+                if lines_since_mismatch >= context_size
+                    && lines_since_mismatch > 0
+                {
+                    results.push(mismatch);
+                    mismatch = Mismatch::new(
+                        line_number - context_queue.len() as u32,
+                        line_number_orig - context_queue.len() as u32,
+                    );
+                }
+
+                while let Some(line) = context_queue.pop_front() {
+                    mismatch
+                        .lines
+                        .push(DiffLine::Context(line.to_owned()));
+                }
```

Comment:
curiousity: would it be possible to create a closure around this code to take the `str` arg and do the stuff here?

Reply:
let me try to make it a little simpler